File: src/main/java/com/zuminX/utils/builder/ApiGenerator.java
package com.zuminX.utils.builder;

import com.intellij.psi.PsiClass;
import com.zuminX.annotations.swagger.Api;
import com.zuminX.interceptor.AnnotationGeneratorInterceptor;
import com.zuminX.utils.GeneratorUtils;
import java.util.Collections;
import java.util.List;

/**
 * Api注解的生成器类
 */
public class ApiGenerator implements AnnotationGenerator<PsiClass, Api> {

  public static final ApiGenerator INSTANCE = AnnotationGeneratorInterceptor.create(ApiGenerator.class, Api.class);

  /**
   * 根据psi类对象构建Api注解对象
   *
   * @param psiClass psi类
   * @return Api注解对象
   */
  public final Api build(PsiClass psiClass) {
    return Api.builder().value(getValue(psiClass)).tags(getTags(psiClass)).build();
  }

  /**
   * 从psi类对象中获取Api类的value属性
   *
   * @param psiClass psi类
   * @return Api类的value属性
   */
  protected String getValue(PsiClass psiClass) {
    return GeneratorUtils.getValueOfRequestMapping(psiClass);
  }

  /**
   * 从psi类对象中获取Api类的tags属性
   *
   * @param psiClass psi类
   * @return Api类的tags属性
   */
  protected List<String> getTags(PsiClass psiClass) {
    String comment = GeneratorUtils.getFirstComment(psiClass);
    return comment == null ? null : Collections.singletonList(comment);
  }
}

File: src/main/java/com/zuminX/generator/builder/ApiGenerator.kt
package com.zuminX.generator.builder

import com.intellij.psi.PsiClass
import com.zuminX.annotations.swagger.Api
import com.zuminX.generator.AnnotationGenerator
import com.zuminX.generator.AnnotationGeneratorInterceptor
import com.zuminX.utils.getFirstComment
import com.zuminX.utils.getValueOfRequestMapping

val ApiGeneratorInstance = AnnotationGeneratorInterceptor.create(ApiGenerator::class.java, Api::class.java)

/**
 * Api注解的生成器类
 */
open class ApiGenerator : AnnotationGenerator<PsiClass, Api> {

  /**
   * 根据psi类对象构建Api注解对象
   *
   * @param psiElement psi类
   * @return Api注解对象
   */
  final override fun build(psiElement: PsiClass): Api {
    return Api().apply {
      value = getValue(psiElement)
      tags = getTags(psiElement)
    }
  }

  /**
   * 从psi类对象中获取Api类的value属性
   *
   * @param psiClass psi类
   * @return Api类的value属性
   */
  protected fun getValue(psiClass: PsiClass): String? {
    return psiClass.getValueOfRequestMapping()
  }

  /**
   * 从psi类对象中获取Api类的tags属性
   *
   * @param psiClass psi类
   * @return Api类的tags属性
   */
  protected fun getTags(psiClass: PsiClass): List<String>? {
    val comment = psiClass.getFirstComment()
    return comment?.let { listOf(it) }
  }
}

